common/client/node_lifecycle.go

```diff
@@ -322,46 +413,114 @@ func (n *node[CHAIN_ID, HEAD, RPC]) outOfSyncLoop(isOutOfSync func(num int64, td
 		return
 	}
 
-	lggr.Tracew("Successfully subscribed to heads feed on out-of-sync RPC node", "nodeState", n.getCachedState())
-
-	ch := make(chan HEAD)
-	sub, err := n.rpc.SubscribeNewHead(ctx, ch)
+	noNewHeadsTimeoutThreshold := n.chainCfg.NodeNoNewHeadsThreshold()
+	headsSub, err := n.registerNewSubscription(ctx, lggr.With("subscriptionType", "heads"),
+		noNewHeadsTimeoutThreshold, n.rpc.SubscribeToHeads)
 	if err != nil {
-		lggr.Errorw("Failed to subscribe heads on out-of-sync RPC node", "nodeState", n.getCachedState(), "err", err)
+		lggr.Errorw("Failed to subscribe heads on out-of-sync RPC node", "err", err)
 		n.declareUnreachable()
 		return
 	}
-	defer sub.Unsubscribe()
 
+	lggr.Tracew("Successfully subscribed to heads feed on out-of-sync RPC node")
+	defer headsSub.Unsubscribe()
+
+	noNewFinalizedBlocksTimeoutThreshold := n.chainCfg.NoNewFinalizedHeadsThreshold()
+	var finalizedHeadsSub headSubscription[HEAD]
+	if n.chainCfg.FinalityTagEnabled() {
+		finalizedHeadsSub, err = n.registerNewSubscription(ctx, lggr.With("subscriptionType", "finalizedHeads"),
+			noNewFinalizedBlocksTimeoutThreshold, n.rpc.SubscribeToFinalizedHeads)
+		if err != nil {
+			lggr.Errorw("Subscribe to finalized heads failed on out-of-sync RPC node", "err", err)
+			n.declareUnreachable()
+			return
+		}
+
+		lggr.Tracew("Successfully subscribed to finalized heads feed on out-of-sync RPC node")
+		defer finalizedHeadsSub.Unsubscribe()
+	}
+
+	_, localHighestChainInfo := n.rpc.GetInterceptedChainInfo()
 	for {
+		if syncIssues == 0 {
+			// back in-sync! flip back into alive loop
+			lggr.Infow(fmt.Sprintf("%s: %s. Node was out-of-sync for %s", msgInSync, n.String(), time.Since(outOfSyncAt)))
+			n.declareInSync()
+			return
+		}
+
 		select {
 		case <-ctx.Done():
 			return
-		case head, open := <-ch:
+		case head, open := <-headsSub.Heads:
 			if !open {
-				lggr.Error("Subscription channel unexpectedly closed", "nodeState", n.getCachedState())
+				lggr.Errorw("Subscription channel unexpectedly closed", "nodeState", n.getCachedState())
 				n.declareUnreachable()
 				return
 			}
-			if !isOutOfSync(head.BlockNumber(), head.BlockDifficulty()) {
-				// back in-sync! flip back into alive loop
-				lggr.Infow(fmt.Sprintf("%s: %s. Node was out-of-sync for %s", msgInSync, n.String(), time.Since(outOfSyncAt)), "blockNumber", head.BlockNumber(), "blockDifficulty", head.BlockDifficulty(), "nodeState", n.getCachedState())
-				n.declareInSync()
-				return
+
+			if !n.onNewHead(lggr, &localHighestChainInfo, head) {
+				continue
+			}
+
```

Comment:
Can you add a one-line comment here to quickly explain how we use bitwise operators? i.e. "We clear NoNewHead sync issue, after receiving a new head and check for out of sync." or something along these lines. It took me some time to figure out what is going on, without context.